### PR TITLE
Fix channel checkpoint issues and fix GC

### PR DESCRIPTION
**Pull request by soothing-rain**
Signed-off-by: Yuchen Gao <yuchen.gao@zilliz.com>

/kind bug

issue: #21897

### CODE REVIEW

File: internal/datacoord/garbage_collector.go [hunk 1]
```diff
@@ -242,25 +242,40 @@ func (gc *garbageCollector) clearEtcd() {
...
-				zap.Uint64("channelCpTs", channelCPs[segment.GetInsertChannel()]),
+				zap.Uint64("channelCpTs", channelCPs[segInsertChannel]),
 			)
 			continue
 		}
 		// For compact A, B -> C, don't GC A or B if C is not indexed,
 		// guarantee replacing A, B with C won't downgrade performance
 		if to, ok := compactTo[segment.GetID()]; ok && !indexedSet.Contain(to.GetID()) {
+			log.Warn("skipping GC when compact target segment is not indexed",
+				zap.Int64("segmentID", to.GetID()))
```
Comment (congqixia) [inline internal/datacoord/garbage_collector.go:261]: Need rated log here

File: internal/datacoord/handler.go [hunk 2]
```diff
@@ -331,7 +331,7 @@ func (h *ServerHandler) CheckShouldDropChannel(channel string) bool {
 			}
 		}
 		return false*/
-	return h.s.meta.catalog.IsChannelDropped(h.s.ctx, channel)
+	return !h.s.meta.catalog.IsChannelDropped(h.s.ctx, channel)
```
Comment (congqixia) [inline internal/datacoord/handler.go:334]: Shall not revert here. IsChannelDropped is determined by an etcd flag for any transaction failure

Comment (soothing-rain) [reply internal/datacoord/handler.go:334]: Yes but I have flipped the method's return values

Comment (congqixia) [reply internal/datacoord/handler.go:334]: Then the method does not behave like expected
https://github.com/milvus-io/milvus/blob/403f85a95fbe60340ae7281ab8161347a8fb130e/internal/metastore/kv/datacoord/kv_catalog.go#L458-L465
Original code means whether a tombstone is marked for this channel. This tombstone was for batch save segment info and mark channel dropped.

Comment (soothing-rain) [reply internal/datacoord/handler.go]: Per offline discussed, we decided to rename `IsChannelDropped` as `ShouldDropChannel` to improve readability.

File: internal/datacoord/garbage_collector.go [hunk 3]
```diff
@@ -242,25 +242,42 @@ func (gc *garbageCollector) clearEtcd() {
...
-		// segment gc shall only happen when channel cp is after segment dml cp.
-		if segment.GetDmlPosition().GetTimestamp() > channelCPs[segment.GetInsertChannel()] {
-			log.WithRateGroup("GC_FAIL_CP_BEFORE", 1, 60).RatedInfo(60, "dropped segment dml position after channel cp, skip meta gc",
-				zap.Uint64("dmlPosTs", segment.GetDmlPosition().GetTimestamp()),
-				zap.Uint64("channelCpTs", channelCPs[segment.GetInsertChannel()]),
-			)
+		segInsertChannel := segment.GetInsertChannel()
+		// Ignore segments from potentially dropped collection. Check if collection is to be dropped by checking if channel is dropped.
+		// We do this because collection meta drop relies on all segment being GCed.
+		if gc.meta.catalog.ShouldDropChannel(context.Background(), segInsertChannel) &&
```
Comment (congqixia) [inline internal/datacoord/garbage_collector.go:248]: maybe logic shall be `!ShouldDropChannel`

**Comment by soothing-rain** [comment]
rerun ut

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**